arangod/GeneralServer/H2CommTask.cpp

```diff
@@ -0,0 +1,654 @@
+        const char* src = body.data() + strm->responseOffset;
+        size_t len = std::min(length, body.size() - strm->responseOffset);
+        TRI_ASSERT(len > 0);
+        std::copy_n(src, len, buf);
+
+        strm->responseOffset += len;
+        if (strm->responseOffset == body.size()) {
+          *data_flags |= NGHTTP2_DATA_FLAG_EOF;
+        }
+
+        return static_cast<ssize_t>(len);
+      };
+      prd_ptr = &prd;
+    }
+
+    int rv = nghttp2_submit_response(this->_session, streamId, nva.data(),
+                                     nva.size(), prd_ptr);
+    if (rv != 0) {
+      this->close();
+      return;
+    }
+  }
+}
+
+// called on IO context thread
+template <SocketType T>
+void H2CommTask<T>::doWrite() {
+  if (_writing) {
+    return;
+  }
+  _writing = true;
+
+  queueHttp2Responses();
+
+  std::array<asio_ns::const_buffer, 2> outBuffers;
+
+  size_t len = 0;
+  while (true) {
+    const uint8_t* data;
+    ssize_t nread = nghttp2_session_mem_send(_session, &data);
+    if (nread < 0) {  // error
+      this->close();
+      return;
+    }
+    if (nread == 0) {  // done
+      break;
+    }
+
+    // if the data is long we just pass it to async_write
+    if (len + nread > _outbuffer.size()) {
+      outBuffers[1] = asio_ns::buffer(data, nread);
+      break;
+    }
+
+    std::copy_n(data, nread, std::begin(_outbuffer) + len);
+    len += nread;
+  }
+  outBuffers[0] = asio_ns::buffer(_outbuffer, len);
+
+  if (asio_ns::buffer_size(outBuffers) == 0) {
+    if (shouldStop()) {
+      this->close();
+    }
+    this->_writing = false;
+    return;
+  }
+
+  // Reset read timer here, because normally client is sending
+  // something, it does not expect timeout while doing it.
+  this->setTimeout(this->DefaultTimeout);
+
+  asio_ns::async_write(this->_protocol->socket, outBuffers,
+                       [self = this->shared_from_this()](const asio_ns::error_code& ec,
+                                                         std::size_t) {
+                         auto& me = static_cast<H2CommTask<T>&>(*self);
+                         if (ec) {
+                           me.close(ec);
+                           return;
+                         }
+
+                         me._writing = false;
+                         me.doWrite();
+                       });
+}
+
+template <SocketType T>
+std::unique_ptr<GeneralResponse> H2CommTask<T>::createResponse(rest::ResponseCode responseCode,
+                                                               uint64_t mid) {
+  return std::make_unique<HttpResponse>(responseCode, mid);
+}
+
+template <SocketType T>
+typename H2CommTask<T>::Stream* H2CommTask<T>::createStream(int32_t sid,
+                                                            std::unique_ptr<HttpRequest> req) {
+  TRI_ASSERT(static_cast<uint64_t>(sid) == req->messageId());
+  auto stream = std::make_unique<Stream>(std::move(req));
+  auto [it, inserted] = _streams.emplace(sid, std::move(stream));
```

Comment:
`inserted` is not used here. this may make compilers unhappy.

Reply:
I thinkt the spec explicitly allows this

Reply:
I have seen at least one compiler cry about this. I think it was g++7, so maybe we don't need to pay much attention to it.